text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -121,6 +120,44 @@ intermediary MAY propagate a transport context to a consumer.
 
 ### Span structure, names, and attributes
 
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+efforts from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Receive"
+spans to correlate producer with consumer traces. A "Receive" spans must link
+to the "Create" spans of all messages that are handled by the respective
+"Receive" operation. Depending on the use case, "Receive" spans can correlate
+with "Process" spans or other spans modelling processing operations.
+
+The operation modelled by the "Receive" span does not strictly refer to
+receiving the message from intermediaries, but instead refers to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Receive" span. However, clients might pre-fetch messages
+from intermediaries and cache those messages, and only forward messages to the
+application at a later time. In this case, the operation of pre-fetching and
+caching should not be covered by the "Receive" span.
+
```

Comment:
If we're saying we can not create "Process" spans because it's not covered by messaging libraries and SDKs, doesn't a "Receive" span become effectively the same as a "Process" in the case of pre-fetching messages for later forwarding to an application?

How do we know an application is going to process a message in the "Receive" case but not the "Process" one?

Reply:
"Process" spans should ideally be created, but we cannot generally require it as there situations where it's not possible to create "Process" spans.

I think "Receive" spans don't become the same as "Process" spans. Although I admit that "Receive" is probably not the best name for those spans, I'm open to other suggestions. I tried to come up with a drawing to clarify what's covered by this span:

![Unbenannte Zeichnung](https://user-images.githubusercontent.com/16337442/146103932-2b87ca3a-f3f7-4543-bef4-f964cae400a3.jpg)

Other spans of the receive stage should be covered by instrumenting the transport layer.

Regarding the relation of the "Receive" span to "Process" spans, we cannot really make hard requirements. There can be a parent/child relationship in some cases, there also can be a link relationship .

Reply:
I think we will end up basically with guidance on both things (receive,process span) and hope that instrumentation/apps follow?

Reply:
I hope this gets clarified by the differentiation introduced between push and pull based instrumentation. If not, please re-open this conversation.